jni/src/Foreign/JNI.hs

```diff
@@ -197,11 +203,28 @@ instance Exception ArrayCopyFailed
 -- | Map Java exceptions to Haskell exceptions.
 throwIfException :: Ptr JNIEnv -> IO a -> IO a
 throwIfException env m = m `finally` do
-    J excptr <- [CU.exp| jthrowable { (*$(JNIEnv *env))->ExceptionOccurred($(JNIEnv *env)) } |]
+    excptr <- [CU.exp| jthrowable { (*$(JNIEnv *env))->ExceptionOccurred($(JNIEnv *env)) } |]
     unless (excptr == nullPtr) $ do
       [CU.exp| void { (*$(JNIEnv *env))->ExceptionDescribe($(JNIEnv *env)) } |]
       [CU.exp| void { (*$(JNIEnv *env))->ExceptionClear($(JNIEnv *env)) } |]
-      throwIO $ JavaException (J excptr)
+      newGlobalJ env excptr >>= throwIO . JavaException
+
+-- | Creates a global reference to a given Java value.
+--
+-- Arranges for 'deleteGlobalRef' to be called when the global reference
+-- is no longer reachable on the Haskell side.
+newGlobalJ :: Ptr JNIEnv -> Ptr (J ty) -> IO (J ty)
+newGlobalJ env (castPtr -> obj) = do
+    g <- [CU.exp| jobject {
+      (*$(JNIEnv *env))->NewGlobalRef($(JNIEnv *env),
+                                      $(jobject obj)) } |]
+    fmap (unsafeCast . J) $ fixIO $ newConcForeignPtr g . finalize
+  where
+    finalize fp = withJNIEnv $ \env ->
+      withForeignPtr fp $ \objp ->
+      [CU.block| void {
+        (*$(JNIEnv *env))->DeleteGlobalRef($(JNIEnv *env),
+                                           $(jobject objp)); } |]
```

Comment:
This has potential to crash the program if run after the jvm has been stopped. It should be made a noop if so.